engines/e_pkcs11.c

```diff
@@ -0,0 +1,771 @@
+   returnType __declspec(dllimport) name
+#else
+#define DEBUG
+#define CK_DECLARE_FUNCTION(returnType, name) \
+   returnType name
+#endif
+
+#ifdef _WIN32
+#define CK_DECLARE_FUNCTION_POINTER(returnType, name) \
+   returnType __declspec(dllimport) (* name)
+#else
+#define CK_DECLARE_FUNCTION_POINTER(returnType, name) \
+   returnType (* name)
+#endif
+
+#define CK_CALLBACK_FUNCTION(returnType, name) \
+   returnType (* name)
+
+#ifndef NULL_PTR
+#define NULL_PTR 0
+#endif
+
+#include "pkcs11.h"
+
+#define PKCS11_CMD_MODULE_PATH            ENGINE_CMD_BASE
+#define PKCS11_CMD_PIN                    (ENGINE_CMD_BASE + 1)
+
+static const ENGINE_CMD_DEFN pkcs11_cmd_defns[] = {
+    {PKCS11_CMD_MODULE_PATH,
+     "MODULE_PATH",
+     "Module path",
+     ENGINE_CMD_FLAG_STRING},
+    {PKCS11_CMD_PIN,
+     "PIN",
+     "PIN",
+     ENGINE_CMD_FLAG_STRING},
+    {0, NULL, NULL, 0}
+};
+
+struct PKCS11_CTX_st {
+    CK_BYTE *id;
+    CK_BYTE *pin;
+    CK_SLOT_ID slotid;
+    CK_SESSION_HANDLE session;
+    CK_OBJECT_HANDLE key;
+    char *module_path;
+};
+
+typedef struct PKCS11_CTX_st PKCS11_CTX;
+static CK_RV pkcs11_initialize(const char *library_path);
+static void pkcs11_finalize(void);
+static CK_SESSION_HANDLE pkcs11_start_session(CK_SLOT_ID slotId);
+static void pkcs11_end_session(CK_SESSION_HANDLE session);
+static int pkcs11_login(PKCS11_CTX *ctx, CK_USER_TYPE userType);
+static int pkcs11_logout(CK_SESSION_HANDLE session);
+static CK_RV pkcs11_load_functions(const char *library_path);
+static int pkcs11_parse_uri(const char *path, char *token, char **value);
+static char pkcs11_hex_int(char nib1, char nib2);
+static int pkcs11_rsa_enc(int flen, const unsigned char *from,
+                          unsigned char *to, RSA *rsa, int padding);
+static RSA_METHOD *pkcs11_rsa(void);
+static PKCS11_CTX *pkcs11_ctx_new(void);
+static void pkcs11_ctx_free(PKCS11_CTX *ctx);
+static CK_SLOT_ID pkcs11_get_slot(PKCS11_CTX *ctx);
+static CK_OBJECT_HANDLE pkcs11_get_private_key(PKCS11_CTX *ctx);
+static CK_FUNCTION_LIST *pkcs11_funcs;
+static void PKCS11_trace(char *format, ...);
+
+/* Engine stuff */
+static int pkcs11_init(ENGINE *e);
+static int pkcs11_bind(ENGINE *e, const char *id);
+static int pkcs11_destroy(ENGINE *e);
+static int pkcs11_finish(ENGINE *e);
+static EVP_PKEY *pkcs11_engine_load_private_key(ENGINE * e, const char *path,
+                                                UI_METHOD * ui_method,
+                                                void *callback_data);
+
+static const char *engine_id = "pkcs11";
+static const char *engine_name = "A minimal PKCS#11 engine only for sign";
+static int pkcs11_idx = -1;
+
+typedef CK_RV pkcs11_pFunc(CK_FUNCTION_LIST **pkcs11_funcs);
+
+static int pkcs11_init(ENGINE *e)
+{
+    PKCS11_CTX *ctx;
+
+    if (pkcs11_idx < 0) {
+        pkcs11_idx = ENGINE_get_ex_new_index(0, NULL, NULL, NULL, 0);
+        if (pkcs11_idx < 0)
+            goto memerr;
+
+        ctx = pkcs11_ctx_new();
```

Comment:
better test it outside this  `if (pkcs11_idx < 0) {` by calling `ctx = ENGINE_get_ex_data(e, pkcs11_idx);`

Reply:
your suggestion causes a malfunction, I have to investigate, for now I leave the previous code

Reply:
I was tinking to something like this :
```C
    if (pkcs11_idx < 0) {
        pkcs11_idx = ENGINE_get_ex_new_index(0, NULL, NULL, NULL, 0);
        if (pkcs11_idx < 0)
            goto memerr;
    }

    ctx = ENGINE_get_ex_data(e, pkcs11_idx);
    if ( ctx == NULL) {
        ctx = pkcs11_ctx_new();
	
        if (ctx == NULL)
            goto memerr;

        ENGINE_set_ex_data(e, pkcs11_idx, ctx);
    }
```

Reply:
ok done, thanks